File: CSharp/Legacy89DiskKit.Application/FileSystem/DirectoryLayoutService.cs
using System.Security.Cryptography;
using System.Text;
using Legacy89DiskKit.Application.CharacterEncoding;
using Legacy89DiskKit.Domain.CharacterEncoding.Interface;
using Legacy89DiskKit.Domain.FileSystem.Interface.FileSystem;
using Legacy89DiskKit.Domain.FileSystem.Interface.Layout;
using Legacy89DiskKit.Domain.FileSystem.Model;
using Legacy89DiskKit.Infrastructure.CharacterEncoding.Encoder;
using Legacy89DiskKit.FileSystem.Application;

namespace Legacy89DiskKit.Application.FileSystem;
public class DirectoryLayoutService
{
    private readonly EncoderRegistry _encoderRegistry;
    public DirectoryLayoutService()
    {
        _encoderRegistry = new EncoderRegistry();
        _encoderRegistry.Register("X1", new X1CharacterEncoder());
        _encoderRegistry.Register("SJIS", new ShiftJisCharacterEncoder());
        _encoderRegistry.Register("Shift-JIS", new ShiftJisCharacterEncoder());
        _encoderRegistry.Register("MSX", new ShiftJisCharacterEncoder());
        _encoderRegistry.Register("PC88", new ShiftJisCharacterEncoder());
    }

    public DirectoryEntryLayout GetLayout(IFileSystem fileSystem)
    {
        return GetProvider(fileSystem).ReadDirectoryLayout();
    }

    public string ExportPlan(IFileSystem fileSystem)
    {
        var layout = GetLayout(fileSystem);
        var lines = new List<string>(layout.Items.Count);
        foreach (var item in layout.Items.OrderBy(item => item.Order))
        {
            if (item.Kind == DirectoryLayoutItemKind.VirtualLabel)
            {
                lines.Add($"# {BuildLabelText(item)}");
                continue;
            }

            lines.Add($"{CreateStableId(item.Id)} {item.DisplayName}");
        }

        return string.Join(Environment.NewLine, lines) + Environment.NewLine;
    }

    public DirectoryLayoutValidationResult ValidatePlan(IFileSystem fileSystem, string text)
    {
        var layout = GetLayout(fileSystem);
        var plan = ParsePlan(text);
        return ValidatePlan(fileSystem, layout, plan);
    }

    public DirectoryLayoutValidationResult ApplyPlan(IFileSystem fileSystem, string text, bool strict = false)
    {
        var provider = GetProvider(fileSystem);
        var validation = ValidatePlan(fileSystem, text);
        if (!validation.IsValid || (strict && validation.WarningCount > 0))
        {
            return validation;
        }

        if (validation.ProposedLayout == null)
        {
            throw new InvalidOperationException("No proposed layout is available.");
        }

        provider.ApplyDirectoryLayout(validation.ProposedLayout);
        return validation;
    }

    public DirectoryEntryLayout MoveEntryBefore(IFileSystem fileSystem, string sourceName, string targetName)
    {
        var provider = GetProvider(fileSystem);
        var layout = provider.ReadDirectoryLayout();
        var source = FindByDisplayName(layout, sourceName);
        var target = FindByDisplayName(layout, targetName);
        var items = layout.Items.ToList();
        items.Remove(source);
        var targetIndex = items.FindIndex(item => item.Id == target.Id);
        items.Insert(targetIndex, source);
        var updated = Reindex(layout, items);
        provider.ApplyDirectoryLayout(updated);
        return updated;
    }

    public DirectoryEntryLayout InsertLabelBefore(IFileSystem fileSystem, string labelText, string targetName)
    {
        var provider = GetProvider(fileSystem);
        var layout = provider.ReadDirectoryLayout();
        var target = FindByDisplayName(layout, targetName);
        var items = layout.Items.ToList();
        var targetIndex = items.FindIndex(item => item.Id == target.Id);
        var label = CreateLabel(fileSystem, labelText, items.Count);
        items.Insert(targetIndex, label);
        var updated = Reindex(layout, items);
        provider.ApplyDirectoryLayout(updated);
        return updated;
    }

    public DirectoryEntryLayout SortEntries(IFileSystem fileSystem, DirectorySortBy sortBy)
    {
        var provider = GetProvider(fileSystem);
        var layout = provider.ReadDirectoryLayout();
        var virtualPositions = layout.Items.Select((item, index) => new { item, index }).Where(x => x.item.Kind == DirectoryLayoutItemKind.VirtualLabel).ToDictionary(x => x.index, x => x.item);
        var sortedFiles = layout.Items.Where(item => item.Kind == DirectoryLayoutItemKind.FileEntry).OrderBy(item => GetSortKey(item, sortBy), StringComparer.OrdinalIgnoreCase).ThenBy(item => item.DisplayName, StringComparer.OrdinalIgnoreCase).ToList();
        var rebuilt = new List<DirectoryLayoutItem>(layout.Items.Count);
        var fileIndex = 0;
        for (var i = 0; i < layout.Items.Count; i++)
        {
            if (virtualPositions.TryGetValue(i, out var virtualItem))
            {
                rebuilt.Add(virtualItem);
                continue;
            }

            rebuilt.Add(sortedFiles[fileIndex++]);
        }

        var updated = Reindex(layout, rebuilt);
        provider.ApplyDirectoryLayout(updated);
        return updated;
    }

    private DirectoryLayoutValidationResult ValidatePlan(IFileSystem fileSystem, DirectoryEntryLayout layout, DirectoryLayoutTextPlan plan)
    {
        var messages = new List<DirectoryLayoutValidationMessage>();
        var fsInfo = fileSystem.GetFileSystemInfo();
        var supportsLayout = fileSystem is IDirectoryLayoutProvider;
        if (!supportsLayout)
        {
            messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, 0, "Directory layout is not supported for this file system."));
            return new DirectoryLayoutValidationResult(plan, messages, null);
        }

        var encoder = ResolveEncoder(fsInfo);
        var filesByStableId = layout.Items.Where(item => item.Kind == DirectoryLayoutItemKind.FileEntry).ToDictionary(item => CreateStableId(item.Id), item => item, StringComparer.OrdinalIgnoreCase);
        var seenIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var usedNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var proposedItems = new List<DirectoryLayoutItem>();
        var consumedFileIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var entry in plan.Entries)
        {
            if (entry.IsLabel)
            {
                DirectoryLayoutItem label;
                try
                {
                    label = CreateLabel(fileSystem, entry.Text, proposedItems.Count);
                }
                catch (Exception ex)
                {
                    messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, entry.LineNumber, ex.Message));
                    continue;
                }

                proposedItems.Add(label);
                continue;
            }

            if (!seenIds.Add(entry.StableId))
            {
                messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, entry.LineNumber, $"Duplicate entry id: {entry.StableId}"));
                continue;
            }

            if (!filesByStableId.TryGetValue(entry.StableId, out var item))
            {
                messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, entry.LineNumber, $"Unknown entry id: {entry.StableId}"));
                continue;
            }

            consumedFileIds.Add(entry.StableId);
            var renamedEntry = ApplyDisplayName(item.Entry!, entry.Text, fsInfo, encoder, entry.LineNumber, messages);
            if (renamedEntry == null)
            {
                continue;
            }

            if (!usedNames.Add(renamedEntry.FullName))
            {
                messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, entry.LineNumber, $"Duplicate target file name: {renamedEntry.FullName}"));
                continue;
            }

            proposedItems.Add(item with { DisplayName = renamedEntry.FullName, Entry = renamedEntry });
        }

        foreach (var item in layout.Items.Where(item => item.Kind == DirectoryLayoutItemKind.FileEntry))
        {
            var stableId = CreateStableId(item.Id);
            if (consumedFileIds.Contains(stableId))
            {
                continue;
            }

            messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Warning, 0, $"Entry omitted from plan and moved to the end: {item.DisplayName}"));
            if (!usedNames.Add(item.DisplayName))
            {
                messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, 0, $"Duplicate target file name: {item.DisplayName}"));
                continue;
            }

            proposedItems.Add(item);
        }

        var capacity = GetDirectoryCapacity(layout, fileSystem);
        if (proposedItems.Count >= capacity)
        {
            messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, 0, $"Directory layout exceeds capacity: {proposedItems.Count}/{capacity - 1}"));
        }

        var proposedLayout = messages.Any(message => message.Severity == DirectoryLayoutValidationSeverity.Error) ? null : Reindex(layout, proposedItems);
        return new DirectoryLayoutValidationResult(plan, messages, proposedLayout);
    }

    private DirectoryLayoutTextPlan ParsePlan(string text)
    {
        var entries = new List<DirectoryLayoutTextPlanEntry>();
        var lines = text.Replace("\r\n", "\n").Split('\n');
        for (var i = 0; i < lines.Length; i++)
        {
            var lineNumber = i + 1;
            var line = lines[i];
            if (string.IsNullOrWhiteSpace(line))
            {
                continue;
            }

            if (line.StartsWith("#"))
            {
                var label = line[1..].Trim();
                entries.Add(new DirectoryLayoutTextPlanEntry(lineNumber, true, string.Empty, label));
                continue;
            }

            var firstWhitespace = line.IndexOfAny([' ', '\t']);
            if (firstWhitespace < 0)
            {
                entries.Add(new DirectoryLayoutTextPlanEntry(lineNumber, false, line.Trim(), string.Empty));
                continue;
            }

            var stableId = line[..firstWhitespace].Trim();
            var displayName = line[(firstWhitespace + 1)..].Trim();
            entries.Add(new DirectoryLayoutTextPlanEntry(lineNumber, false, stableId, displayName));
        }

        return new DirectoryLayoutTextPlan(entries);
    }

    private static IDirectoryLayoutProvider GetProvider(IFileSystem fileSystem)
    {
        if (fileSystem is IDirectoryLayoutProvider provider)
        {
            return provider;
        }

        throw new InvalidOperationException("Directory layout is not supported for this file system.");
    }

    private static DirectoryLayoutItem FindByDisplayName(DirectoryEntryLayout layout, string displayName)
    {
        var item = layout.Items.FirstOrDefault(candidate => candidate.DisplayName.Equals(displayName, StringComparison.OrdinalIgnoreCase));
        if (item == null)
        {
            throw new InvalidOperationException($"Directory entry not found: {displayName}");
        }

        return item;
    }

    private static DirectoryEntryLayout Reindex(DirectoryEntryLayout layout, IReadOnlyList<DirectoryLayoutItem> items)
    {
        return layout with
        {
            Items = items.Select((item, index) => item with { Order = index }).ToArray()
        };
    }

    private static string GetSortKey(DirectoryLayoutItem item, DirectorySortBy sortBy)
    {
        return sortBy switch
        {
            DirectorySortBy.Extension => item.Entry?.Extension ?? item.VirtualLabel?.Extension ?? string.Empty,
            DirectorySortBy.Type => item.Entry?.FileSystemMetadata is HuBasicFileMetadata metadata ? metadata.FileType.ToString() : item.Kind.ToString(),
            _ => item.Entry?.FileName ?? item.VirtualLabel?.FileName ?? item.DisplayName
        };
    }

    private FileEntry? ApplyDisplayName(FileEntry entry, string displayName, DiskFileSystemInfo fsInfo, ICharacterEncoder encoder, int lineNumber, List<DirectoryLayoutValidationMessage> messages)
    {
        if (string.IsNullOrWhiteSpace(displayName))
        {
            messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, lineNumber, "File name is required."));
            return null;
        }

        var parts = SplitDisplayName(NormalizeHostText(displayName), fsInfo);
        if (parts == null)
        {
            messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, lineNumber, $"Invalid file name: {displayName}"));
            return null;
        }

        if (!ValidateNameForFileSystem(parts.Value.Name, parts.Value.Extension, fsInfo, encoder, lineNumber, messages))
        {
            return null;
        }

        byte[]? rawName = entry.RawFileName;
        byte[]? rawExtension = entry.RawExtension;
        if (fsInfo.FileSystemName == "Hu-BASIC")
        {
            rawName = EncodePadded(encoder, parts.Value.Name, 13);
            rawExtension = EncodePadded(encoder, parts.Value.Extension, 3);
        }

        return entry with
        {
            FileName = parts.Value.Name,
            Extension = parts.Value.Extension,
            RawFileName = rawName,
            RawExtension = rawExtension
        };
    }

    private DirectoryLayoutItem CreateLabel(IFileSystem fileSystem, string text, int seed)
    {
        var fsInfo = fileSystem.GetFileSystemInfo();
        if (fsInfo.FileSystemName != "Hu-BASIC")
        {
            throw new InvalidOperationException("Layout labels are only supported for Hu-BASIC.");
        }

        var encoder = ResolveEncoder(fsInfo);
        var label = ParseLabelText(text, fsInfo, encoder);
        var id = $"label:{seed + 1}:{label.FileName}.{label.Extension}";
        return new DirectoryLayoutItem(id, seed, DirectoryLayoutItemKind.VirtualLabel, BuildDisplayName(label.FileName, label.Extension), null, label);
    }

    private static (string Name, string Extension)? SplitDisplayName(string displayName, DiskFileSystemInfo fsInfo)
    {
        var trimmed = displayName.Trim();
        if (string.IsNullOrEmpty(trimmed))
        {
            return null;
        }

        if (fsInfo.FileSystemName == "Hu-BASIC")
        {
            var lastDot = trimmed.LastIndexOf('.');
            if (lastDot <= 0 || lastDot == trimmed.Length - 1)
            {
                return (trimmed, string.Empty);
            }

            var extension = trimmed[(lastDot + 1)..];
            if (extension.Length > 3)
            {
                return (trimmed, string.Empty);
            }

            return (trimmed[..lastDot], extension);
        }

        var dot = trimmed.LastIndexOf('.');
        return dot > 0 ? (trimmed[..dot], trimmed[(dot + 1)..]) : (trimmed, string.Empty);
    }

    private static bool ValidateNameForFileSystem(string fileName, string extension, DiskFileSystemInfo fsInfo, ICharacterEncoder encoder, int lineNumber, List<DirectoryLayoutValidationMessage> messages)
    {
        if (fsInfo.FileSystemName == "Hu-BASIC")
        {
            if (encoder.EncodeText(fileName).Length > 13)
            {
                messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, lineNumber, $"File name exceeds 13 bytes: {fileName}"));
                return false;
            }

            if (encoder.EncodeText(extension).Length > 3)
            {
                messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, lineNumber, $"Extension exceeds 3 bytes: {extension}"));
                return false;
            }

            if (!IsRoundTripSafe(encoder, fileName) || !IsRoundTripSafe(encoder, extension))
            {
                messages.Add(new DirectoryLayoutValidationMessage(DirectoryLayoutValidationSeverity.Error, lineNumber, $"File name contains characters not representable on {fsInfo.PlatformId}: {BuildDisplayName(fileName, extension)}"));
                return false;
            }
        }

        return true;
    }

    private static string NormalizeHostText(string text) => text;
    private static byte[] EncodePadded(ICharacterEncoder encoder, string text, int width)
    {
        var encoded = encoder.EncodeText(text);
        var padded = Enumerable.Repeat((byte)0x20, width).ToArray();
        Array.Copy(encoded, padded, Math.Min(encoded.Length, padded.Length));
        return padded;
    }

    private static bool IsRoundTripSafe(ICharacterEncoder encoder, string text)
    {
        if (string.IsNullOrEmpty(text))
        {
            return true;
        }

        var roundTrip = encoder.DecodeText(encoder.EncodeText(text)).TrimEnd(' ');
        return string.Equals(roundTrip, text, StringComparison.Ordinal);
    }

    private ICharacterEncoder ResolveEncoder(DiskFileSystemInfo fsInfo)
    {
        return new CharacterEncodingResolver(_encoderRegistry).ResolveEncoder(fsInfo);
    }

    private static int GetDirectoryCapacity(DirectoryEntryLayout layout, IFileSystem fileSystem)
    {
        return fileSystem.GetFileSystemInfo().FileSystemName == "Hu-BASIC" ? 128 : layout.Items.Count + 1;
    }

    private static string BuildDisplayName(string fileName, string extension)
    {
        return string.IsNullOrEmpty(extension) ? fileName : $"{fileName}.{extension}";
    }

    private static string BuildLabelText(DirectoryLayoutItem item)
    {
        if (item.VirtualLabel == null)
        {
            return item.DisplayName;
        }

        return BuildDisplayName(item.VirtualLabel.FileName, item.VirtualLabel.Extension);
    }

    private static VirtualDirectoryLabelEntry ParseLabelText(string text, DiskFileSystemInfo fsInfo, ICharacterEncoder encoder)
    {
        var normalizedText = NormalizeHostText(text.Trim());
        if (string.IsNullOrWhiteSpace(normalizedText))
        {
            throw new InvalidOperationException("Label text is required.");
        }

        var firstDot = normalizedText.IndexOf('.');
        string fileName;
        string extension;
        if (firstDot < 0)
        {
            fileName = normalizedText;
            extension = string.Empty;
        }
        else
        {
            fileName = normalizedText[..firstDot];
            var remaining = normalizedText[(firstDot + 1)..];
            var secondDot = remaining.IndexOf('.');
            extension = secondDot >= 0 ? remaining[..secondDot] : remaining;
        }

        fileName = TrimToEncodedBytes(fileName.Trim(), encoder, 13);
        extension = TrimToEncodedBytes(extension.Trim(), encoder, 3);
        if (string.IsNullOrWhiteSpace(fileName))
        {
            throw new InvalidOperationException("Label name is required.");
        }

        if (!ValidateLabelPart(fileName, encoder) || (!string.IsNullOrEmpty(extension) && !ValidateLabelPart(extension, encoder)))
        {
            throw new InvalidOperationException($"Label text contains characters not representable on {fsInfo.PlatformId}: {normalizedText}");
        }

        return new VirtualDirectoryLabelEntry(fileName, extension, 0x44, 0x01, 0, 0xFFFF, 0xFFFE, 0xFFFF, 0x3FFF);
    }

    private static bool ValidateLabelPart(string text, ICharacterEncoder encoder)
    {
        return IsRoundTripSafe(encoder, text);
    }

    private static string TrimToEncodedBytes(string text, ICharacterEncoder encoder, int maxBytes)
    {
        if (string.IsNullOrEmpty(text))
        {
            return string.Empty;
        }

        var value = text;
        while (encoder.EncodeText(value).Length > maxBytes && value.Length > 0)
        {
            value = value[..^1];
        }

        return value;
    }

    public static string CreateStableId(string rawId)
    {
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(rawId));
        return Convert.ToHexString(hash[..4]);
    }
}

File: Cpp/Legacy89DiskKit.Cpp/src/application/directory_layout_service.cpp
#include "legacy89diskkit/cpp/application/directory_layout_service.hpp"
#include <algorithm>
#include <iomanip>
#include <sstream>
#include <set>

namespace legacy89diskkit::cpp::application
{
DirectoryLayoutService::DirectoryLayoutService(NativeFileSystemSession* session)
    : session_(session)
{
}

Result<DirectoryLayout> DirectoryLayoutService::GetLayout() const
{
    if (!session_)
    {
        return Result<DirectoryLayout>::Failure(StatusCode::InvalidArgument, "Native session is not initialized.");
    }

    auto layout_result = session_->ReadDirectoryLayout();
    if (!layout_result.ok())
    {
        return layout_result;
    }

    auto layout = std::move(layout_result.value());
    for (auto& item : layout.items)
    {
        item.stable_id = CreateStableId(item.id);
    }

    return Result<DirectoryLayout>::Success(std::move(layout));
}

Result<std::string> DirectoryLayoutService::ExportPlan() const
{
    auto layout_result = GetLayout();
    if (!layout_result.ok())
    {
        return Result<std::string>::Failure(layout_result.status().code, layout_result.status().message);
    }

    std::stringstream ss;
    auto items = layout_result.value().items;
    std::sort(items.begin(), items.end(), [](const auto& a, const auto& b) { return a.order < b.order; });

    for (const auto& item : items)
    {
        if (item.kind == DirectoryLayoutItemKind::VirtualLabel)
        {
            ss << "# " << item.display_name << "\n";
        }
        else
        {
            ss << item.stable_id << " " << item.display_name << "\n";
        }
    }

    return Result<std::string>::Success(ss.str());
}

Result<DirectoryLayoutValidationResult> DirectoryLayoutService::ValidatePlan(const std::string& plan_text) const
{
    auto current_layout_result = GetLayout();
    if (!current_layout_result.ok())
    {
        return Result<DirectoryLayoutValidationResult>::Failure(current_layout_result.status().code, current_layout_result.status().message);
    }

    const auto& current_layout = current_layout_result.value();
    DirectoryLayoutValidationResult result;
    result.is_valid = true;
    result.warning_count = 0;

    std::vector<DirectoryLayoutItem> proposed_items;
    std::set<std::string> seen_stable_ids;
    std::set<std::string> consumed_ids;

    std::stringstream ss(plan_text);
    std::string line;
    int line_number = 0;

    while (std::getline(ss, line))
    {
        line_number++;
        if (line.empty() || std::all_of(line.begin(), line.end(), isspace)) continue;

        if (line[0] == '#')
        {
            // Label
            DirectoryLayoutItem label;
            label.kind = DirectoryLayoutItemKind::VirtualLabel;
            label.display_name = line.substr(1);
            // Trim leading/trailing spaces from label
            label.display_name.erase(0, label.display_name.find_first_not_of(" "));
            label.display_name.erase(label.display_name.find_last_not_of(" ") + 1);
            label.id = "label:" + std::to_string(proposed_items.size()) + ":" + label.display_name;
            label.stable_id = ""; // Labels don't need stable IDs in the same way
            label.order = static_cast<int>(proposed_items.size());
            proposed_items.push_back(std::move(label));
            continue;
        }

        std::stringstream line_ss(line);
        std::string stable_id, display_name;
        line_ss >> stable_id;
        std::getline(line_ss, display_name);
        if (!display_name.empty() && display_name[0] == ' ') display_name.erase(0, 1);

        if (seen_stable_ids.count(stable_id))
        {
            result.messages.push_back({DirectoryLayoutValidationSeverity::Error, line_number, "Duplicate entry id: " + stable_id});
            result.is_valid = false;
            continue;
        }
        seen_stable_ids.insert(stable_id);

        auto it = std::find_if(current_layout.items.begin(), current_layout.items.end(), 
            [&](const auto& item) { return item.stable_id == stable_id; });

        if (it == current_layout.items.end())
        {
            result.messages.push_back({DirectoryLayoutValidationSeverity::Error, line_number, "Unknown entry id: " + stable_id});
            result.is_valid = false;
            continue;
        }

        consumed_ids.insert(it->id);
        DirectoryLayoutItem proposed = *it;
        proposed.display_name = display_name;
        proposed.order = static_cast<int>(proposed_items.size());
        proposed_items.push_back(std::move(proposed));
    }

    // Check for omitted items
    for (const auto& item : current_layout.items)
    {
        if (item.kind == DirectoryLayoutItemKind::FileEntry && consumed_ids.find(item.id) == consumed_ids.end())
        {
            result.warning_count++;
            result.messages.push_back({DirectoryLayoutValidationSeverity::Warning, 0, "Entry omitted from plan and moved to the end: " + item.display_name});
            DirectoryLayoutItem omitted = item;
            omitted.order = static_cast<int>(proposed_items.size());
            proposed_items.push_back(std::move(omitted));
        }
    }

    if (result.is_valid)
    {
        DirectoryLayout layout;
        layout.items = std::move(proposed_items);
        result.proposed_layout = std::move(layout);
    }

    return Result<DirectoryLayoutValidationResult>::Success(std::move(result));
}

Result<DirectoryLayoutValidationResult> DirectoryLayoutService::ApplyPlan(const std::string& plan_text, bool strict)
{
    auto validation_result = ValidatePlan(plan_text);
    if (!validation_result.ok()) return validation_result;

    auto& validation = validation_result.value();
    if (!validation.is_valid || (strict && validation.warning_count > 0))
    {
        return Result<DirectoryLayoutValidationResult>::Success(std::move(validation));
    }

    if (!validation.proposed_layout.has_value())
    {
        return Result<DirectoryLayoutValidationResult>::Failure(StatusCode::InvalidArgument, "No proposed layout available.");
    }

    auto status = session_->ApplyDirectoryLayout(validation.proposed_layout.value());
    if (!status.ok())
    {
        return Result<DirectoryLayoutValidationResult>::Failure(status.code, status.message);
    }

    return Result<DirectoryLayoutValidationResult>::Success(std::move(validation));
}

Result<DirectoryLayout> DirectoryLayoutService::MoveEntryBefore(const std::string& source_name, const std::string& target_name)
{
    auto layout_result = GetLayout();
    if (!layout_result.ok()) return layout_result;

    auto items = std::move(layout_result.value().items);
    auto source_it = std::find_if(items.begin(), items.end(), [&](const auto& i) { return i.display_name == source_name; });
    auto target_it = std::find_if(items.begin(), items.end(), [&](const auto& i) { return i.display_name == target_name; });

    if (source_it == items.end() || target_it == items.end())
    {
        return Result<DirectoryLayout>::Failure(StatusCode::InvalidArgument, "Entry not found.");
    }

    auto source_item = std::move(*source_it);
    items.erase(source_it);
    
    // Refresh target_it after erase
    target_it = std::find_if(items.begin(), items.end(), [&](const auto& i) { return i.display_name == target_name; });
    items.insert(target_it, std::move(source_item));

    return Reindex(items);
}

Result<DirectoryLayout> DirectoryLayoutService::InsertLabelBefore(const std::string& label_text, const std::string& target_name)
{
    auto layout_result = GetLayout();
    if (!layout_result.ok()) return layout_result;

    auto items = std::move(layout_result.value().items);
    auto target_it = std::find_if(items.begin(), items.end(), [&](const auto& i) { return i.display_name == target_name; });

    if (target_it == items.end())
    {
        return Result<DirectoryLayout>::Failure(StatusCode::InvalidArgument, "Target entry not found.");
    }

    DirectoryLayoutItem label;
    label.kind = DirectoryLayoutItemKind::VirtualLabel;
    label.display_name = label_text;
    label.id = "label:" + std::to_string(std::hash<std::string>{}(label_text)) + ":" + label_text;
    label.stable_id = "";
    
    items.insert(target_it, std::move(label));

    return Reindex(items);
}

Result<DirectoryLayout> DirectoryLayoutService::SortEntries(DirectorySortBy sort_by)
{
    if (sort_by != DirectorySortBy::Name)
    {
        return Result<DirectoryLayout>::Failure(StatusCode::UnsupportedFormat, "Only Name-based sorting is currently supported.");
    }

    auto layout_result = GetLayout();
    if (!layout_result.ok()) return layout_result;

    auto items = std::move(layout_result.value().items);
    
    // Separate labels and files to maintain label positions if possible (like C#)
    std::stable_sort(items.begin(), items.end(), [&](const auto& a, const auto& b) {
        if (a.kind == DirectoryLayoutItemKind::VirtualLabel || b.kind == DirectoryLayoutItemKind::VirtualLabel) return false;
        return a.display_name < b.display_name;
    });

    return Reindex(items);
}

std::string DirectoryLayoutService::CreateStableId(const std::string& raw_id)
{
    // WARNING: std::hash is not guaranteed to be stable across different runs or platforms.
    // This is a temporary measure for Phase V2-25. Full parity with C# SHA256 IDs 
    // should be implemented in a later phase when a crypto provider is available.
    size_t hash = std::hash<std::string>{}(raw_id);
    std::stringstream ss;
    ss << std::hex << std::setw(8) << std::setfill('0') << (hash & 0xFFFFFFFF);
    return ss.str().substr(0, 8);
}

Result<DirectoryLayout> DirectoryLayoutService::Reindex(const std::vector<DirectoryLayoutItem>& items) const
{
    DirectoryLayout layout;
    layout.items = items;
    for (size_t i = 0; i < layout.items.size(); ++i)
    {
        layout.items[i].order = static_cast<int>(i);
    }

    auto status = session_->ApplyDirectoryLayout(layout);
    if (!status.ok())
    {
        return Result<DirectoryLayout>::Failure(status.code, status.message);
    }

    return Result<DirectoryLayout>::Success(std::move(layout));
}
} // namespace legacy89diskkit::cpp::application
